model/src/signing.rs

```diff
@@ -0,0 +1,43 @@
+use crate::{DomainSeparator, Signature};
+use web3::signing;
+use web3::types::Recovery;
+use primitive_types::H160;
+
+pub trait EIP712Signing {
+
+    fn digest(&self) -> [u8; 32];
+
+    fn signing_digest_typed_data(&self, domain_separator: &DomainSeparator) -> [u8; 32] {
+        let mut hash_data = [0u8; 66];
+        hash_data[0..2].copy_from_slice(&[0x19, 0x01]);
+        hash_data[2..34].copy_from_slice(&domain_separator.0);
+        hash_data[34..66].copy_from_slice(&self.digest());
+        signing::keccak256(&hash_data)
+    }
+
+    fn signing_digest_message(&self, domain_separator: &DomainSeparator) -> [u8; 32] {
+        let mut hash_data = [0u8; 92];
+        hash_data[0..28].copy_from_slice(b"\x19Ethereum Signed Message:\n64");
+        hash_data[28..60].copy_from_slice(&domain_separator.0);
+        hash_data[60..92].copy_from_slice(&self.digest());
+        signing::keccak256(&hash_data)
+    }
+
+    fn signing_digest(&self, domain_separator: &DomainSeparator, signature: Signature) -> [u8; 32] {
+        // This is fallback for wallets that don't support EIP712
+        if signature.v & 0x80 == 0 {
+            self.signing_digest_typed_data(domain_separator)
+        } else {
+            self.signing_digest_message(domain_separator)
+        }
+    }
+
+    fn validate_signature(&self, domain_separator: &DomainSeparator, signature: Signature) -> Option<H160> {
```

Comment:
Instead of creating a trait that needs to be added to all structs that carry a signature, I would add these methods to `impl Signature` directly and have it take the digest and domain separator as an arguments. Then expose `validate_signature` on OrderCreation/Cancellation and internally call `self.signature.validate(domain_separator, digest)`

Reply:
Very nice idea. This could clear up the problem with the trait I was having.